### ISSUE TITLE
WindowGate

### ISSUE THREAD

**Issue by nicholas-leonard** [nicholas-leonard/cunnx#12]
This Module is for another Conditional Computation project. It must follow a SoftMax.
It computes the expectation over outputs to get the most active coordinate. So for an output of just 100 softmax categories, we might have a most active coordinate of 43.32113. This coordinate can then be used to get the start index of a window into a axis of much larger dimension, say `maxIndex=1000`. The starting index of the window might then be 433. The window may be of size `windowSize=256`. The downscaled window into the softmax would be `batchSize x input:size(2)*windowSize/maxIndex`.

I am thinking that the output of the module is a table of `{indice, scale}` where `scale` is a matrix of size `batchSize x windowSize`. The scale matrix would be the result of an up-scaling of a narrow window into the softmax output (the input of this module). So the idea would be to backpropagate through those scales. The backpropagation would need  to move the window left or right over the inputs.
### Option 1

In my previous work, I used a gaussian blur as softmax targets for a similar problem. I wonder if I couldn't do something similar here. My target was a facial keypoint coordinate (one softmax per axis). The gaussian blur target would keep the output of the softmax uni-modal. In our case, the target is unknown. Maybe we could measure the expectation over the gradOutputs to get a target coordinate. We can then create a gaussian blur target centred at that target coordinate to force a kind of uni-modality.
### Option 2

While a target gaussian blur makes for a great target, it is difficult to obtain a coordinate target from a window of gradients. The target coordinate would have to locate the area with most promise for the input. So is that the area with the gradients having the highest or lowest amplitude (abs(gradient)) or having the highest gradient, etc?. So maybe we should just forget the gaussian blur approach and directly backpropagate the gradients through the window. It is easier to implement anyway.
## Balance

The only remaining problem would be to keep those targets. In order words, you don't want all targets to be concentrated in one area. This is where we could apply a version of zhen zhou's idea where we modify the biases to balance the probability of categories over all examples.
